Add support for MongoDBContainer credentials

digital88 opened the pull request:
Hopefully closes #988

This implementation supports credentials only when replica set is disabled. Because using credentials with replica set requires rather complicated setup I think it could be implemented later as separate PR. I'm not even sure _why_ replica set was implemented in first place, as TC creates single container for running tests it does not make much sense IMO. Either way the old replica set code remains untouched and is used by default, so nothing should break for existing users.

cristianrgreco commented on packages/modules/mongodb/src/mongodb-container.test.ts:
```diff
@@ -64,4 +64,47 @@ describe("MongodbContainer", { timeout: 240_000 }, () => {
+      const mongodbContainer = await new MongoDBContainer(image)
+        .withUsername("mongo_user")
+        .withPassword("mongo_password")
+        .withReplicaSetEnabled(false)
```
Once we implement credentials support with replica sets, I guess we wouldn’t ever need to disable it. I’m therefore hesitant to add this to the API just to support credentials. WDYT, is it a lot more work to make it work with replica sets? Anything I can help with?

digital88 replied:
Ok, I'll take some more approaches. Will try to make RS + auth possible. I will document what works and what is not.

digital88 commented on packages/modules/mongodb/src/mongodb-container.test.ts:
```diff
@@ -64,4 +64,47 @@ describe("MongodbContainer", { timeout: 240_000 }, () => {
     await mongodbContainer.stop();
   });
   // }
+  test.for([["mongo:4.0.1"], ["mongo:6.0.1"], ["mongo:8.0"]])(
```
Thanks, I will use only 8.0 image.

digital88 commented:
Also very interesting that 4.0 image fails tests with:
```
× MongodbContainer > should connect to mongo:4.0.1 with credentials and disabled replica set 127121ms (retry x3)
     → Server at localhost:41731 reports maximum wire version 7, but this version of the Node.js Driver requires at least 8 (MongoDB 4.2)
```
I do not encounter this on local machine. Need to investigate more.

digital88 commented:
> Also very interesting that 4.0 image fails tests with:

Had to pin mongoose to `"8.15.0"`. Newer versions [do not support](https://mongoosejs.com/docs/compatibility.html) mongo 4.

I moved all logic in `containerStarted` to healthcheck and changed wait strategy to wait for healthcheck. Took inspiration from searching how to run mongo replica set with auth in compose. Not sure about default timeouts tho.

digital88 commented:
Changed `CMD` to `CMD-SHELL` as used in other modules. Tests pass in podman locally.

cristianrgreco commented:
@digital88 do you mind resolving the conflicts? I would but I’m afk at the moment. Should just be a testcontainers version bump

digital88 commented:
> @digital88 do you mind resolving the conflicts? I would but I’m afk at the moment. Should just be a testcontainers version bump

Resolved

digital88 commented:
> > @digital88 do you mind resolving the conflicts? I would but I’m afk at the moment. Should just be a testcontainers version bump
> 
> Resolved

And resolved again 😄

cristianrgreco commented:
Apologies @digital88, things are a bit busy at the moment, I know I need to review this PR, will get to it soon.

cristianrgreco commented on packages/modules/mongodb/src/mongodb-container.ts:
```diff
@@ -1,68 +1,85 @@
 const MONGODB_PORT = 27017;
 
 export class MongoDBContainer extends GenericContainer {
+  private username = "";
+  private password = "";
```
Let's remove the default values and have the type `string | undefined`, so I don't have to double check `if (this.username && this.password)`  works as expected

cristianrgreco commented on packages/modules/mongodb/src/mongodb-container.ts:
```diff
@@ -1,68 +1,85 @@
+    if (useMongosh) args.push("mongosh");
+    else args.push("mongo", "admin");
+    if (this.username && this.password) args.push("-u", this.username, "-p", this.password);
+    args.push("--host", "localhost", "--quiet", "--eval", command);
```
What was the reasoning for setting the host? If Mongo is able to set it itself I would opt for that

digital88 replied:
Leftover when I was experimenting. Removed.

cristianrgreco commented on packages/modules/mongodb/src/mongodb-container.ts:
```diff
@@ -1,68 +1,85 @@
-    const execResult = await startedTestContainer.exec(this.buildMongoEvalCommand(command));
-    this.checkMongoNodeExitCode(execResult);
+  public override async start(): Promise<StartedMongoDBContainer> {
+    const cmdArgs = ["--replSet", "rs0", "--bind_ip_all"];
```
How comes `bind_ip_all` was set, could it be a breaking change for existing users?

digital88 replied:
Leftover when I was experimenting. `bind_ip_all` equals to "bind to 0.0.0.0" inside container. Removed.

digital88 commented on packages/modules/mongodb/src/mongodb-container.ts:
```diff
@@ -1,68 +1,85 @@
-    this.checkMongoNodeExitCode(execResult);
+  public override async start(): Promise<StartedMongoDBContainer> {
+    const cmdArgs = ["--replSet", "rs0", "--bind_ip_all"];
+    this.withHealthCheck({
```
Fixed.

digital88 commented:
Oh, sorry, TC minor just went up again...Will fix it in a minute.

cristianrgreco commented on packages/modules/mongodb/src/mongodb-container.test.ts:
```diff
@@ -63,3 +63,27 @@ describe("MongodbContainer", { timeout: 240_000 }, () => {
 });
+
+describe("MongodbContainer connect with credentials", { timeout: 240_000 }, () => {
+  it("should connect to %s with credentials", async () => {
```
This no longer needs to be in a describe, and the image `mongo:8.0` should be replaced with `IMAGE` (so it'll be auto-upgraded by Dependabot)

digital88 replied:
Ok. I see that test named `should work using default version 4.0.1` seems to use latest image 8.x.x too (It uses IMAGE variable)? Maybe we should hardcode mongo version to 4 for this test or rename this test?

digital88 commented:
> @digital88 Are there any more changes you'd like to make?

I think it's good 👍

cristianrgreco commented:
Thanks for the contribution, always a pleasure!
